// mi6-game-token-manager/index.js
const jwt = require('jsonwebtoken');

class JwtManager {
  constructor(secret) {
    this.secret = secret || 'mi6-secret-key-change-this';
  }

  /**
   * Generate a new JWT token
   * @param {Object} payload - Data to encode
   * @param {String} expiresIn - Expiration time (e.g., '1h', '7d')
   * @returns {String} - The generated token
   */
  generateToken(payload, expiresIn = '24h') {
    return jwt.sign(payload, this.secret, { expiresIn });
  }

  /**
   * Verify a JWT token
   * @param {String} token - The token to verify
   * @returns {Object|null} - Decoded payload or null if invalid
   */
  verifyToken(token) {
    try {
      return jwt.verify(token, this.secret);
    } catch (error) {
      return null;
    }
  }

  /**
   * Decode a token without verifying
   * @param {String} token 
   * @returns {Object}
   */
  decodeToken(token) {
    return jwt.decode(token);
  }

  /**
   * Express Middleware to protect routes
   */
  middleware() {
    return (req, res, next) => {
      const authHeader = req.headers['authorization'];
      const token = authHeader && authHeader.split(' ')[1];

      if (!token) {
        return res.status(401).json({ message: 'Access Denied: No Token Provided' });
      }

      const verified = this.verifyToken(token);
      if (!verified) {
        return res.status(403).json({ message: 'Invalid Token' });
      }

      req.user = verified;
      next();
    };
  }
}

class ApiKeyManager {
  constructor(validKeys) {
    // Use a Set for O(1) lookups and easy addition/removal
    this.validKeys = new Set(Array.isArray(validKeys) ? validKeys : [validKeys]);
  }

  /**
   * Check if a key is valid
   * @param {String} key 
   * @returns {Boolean}
   */
  isValid(key) {
    return this.validKeys.has(key);
  }

  /**
   * Add a new valid API key
   * @param {String} key 
   */
  addKey(key) {
    this.validKeys.add(key);
  }

  /**
   * Remove an API key
   * @param {String} key 
   */
  removeKey(key) {
    this.validKeys.delete(key);
  }

  /**
   * Express Middleware to check API Key
   * Expects header 'x-api-key'
   */
  middleware() {
    return (req, res, next) => {
      const apiKey = req.header('x-api-key');
      if (!apiKey || !this.isValid(apiKey)) {
        return res.status(401).json({ message: 'Invalid API Key' });
      }
      next();
    };
  }
}

module.exports = {
  JwtManager,
  ApiKeyManager
};
